# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/base/WindowContext.h
# Commit: 0a44c410b59b
# Full Hash: 0a44c410b59bbc54bfd80ce598562e4362961d7e
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-04-24 09:51:27
# Regressor Bug: 1580565
# File Overlap Count: 3
# Description:
#   Bug 1580565 - Part 4: Use WindowContext to manage BrowsingContext cached status, r=farre
#   
#   The existing infrastructure which stored cached BrowsingContexts on the
#   BrowsingContextGroup was added before WindowContexts were added, and can cause
#   racing issues with partially discarded trees during process switches.
# ==============================================================================

diff -r 20dbcfc9eacc -r 0a44c410b59b docshell/base/WindowContext.h
--- a/docshell/base/WindowContext.h	Thu Apr 23 21:52:44 2020 +0000
+++ b/docshell/base/WindowContext.h	Thu Apr 23 21:52:47 2020 +0000
@@ -7,6 +7,7 @@
 #ifndef mozilla_dom_WindowContext_h
 #define mozilla_dom_WindowContext_h
 
+#include "mozilla/Span.h"
 #include "mozilla/dom/MaybeDiscarded.h"
 #include "mozilla/dom/SyncedContext.h"
 #include "mozilla/net/NeckoChannelParams.h"
@@ -16,6 +17,8 @@
 
 class WindowGlobalParent;
 
+class WindowGlobalParent;
+
 #define MOZ_EACH_WC_FIELD(FIELD)                                       \
   FIELD(OuterWindowId, uint64_t)                                       \
   FIELD(CookieJarSettings, Maybe<mozilla::net::CookieJarSettingsArgs>) \
@@ -44,6 +47,10 @@
   uint64_t OuterWindowId() const { return GetOuterWindowId(); }
   bool IsDiscarded() const { return mIsDiscarded; }
 
+  bool IsCached() const;
+
+  Span<RefPtr<BrowsingContext>> Children() { return mChildren; }
+
   // Cast this object to it's parent-process canonical form.
   WindowGlobalParent* Canonical();
 
@@ -58,6 +65,12 @@
     uint64_t mBrowsingContextId;
 
     FieldTuple mFields;
+
+    bool operator==(const IPCInitializer& aOther) const {
+      return mInnerWindowId == aOther.mInnerWindowId &&
+             mBrowsingContextId == aOther.mBrowsingContextId &&
+             mFields == aOther.mFields;
+    }
   };
   IPCInitializer GetIPCInitializer() {
     return {mInnerWindowId, mBrowsingContext->Id(), mFields.Fields()};
@@ -74,6 +87,11 @@
   void Init();
 
  private:
+  friend class BrowsingContext;
+
+  void AppendChildBrowsingContext(BrowsingContext* aBrowsingContext);
+  void RemoveChildBrowsingContext(BrowsingContext* aBrowsingContext);
+
   // Send a given `BaseTransaction` object to the correct remote.
   void SendCommitTransaction(ContentParent* aParent,
                              const BaseTransaction& aTxn, uint64_t aEpoch);
@@ -114,6 +132,12 @@
   uint64_t mInnerWindowId;
   RefPtr<BrowsingContext> mBrowsingContext;
 
+  // --- NEVER CHANGE `mChildren` DIRECTLY! ---
+  // Changes to this list need to be synchronized to the list within our
+  // `mBrowsingContext`, and should only be performed through the
+  // `AppendChildBrowsingContext` and `RemoveChildBrowsingContext` methods.
+  nsTArray<RefPtr<BrowsingContext>> mChildren;
+
   bool mIsDiscarded = false;
 };
 